DigitalLearningSolutions.Web/Controllers/LoginController.cs

```diff
@@ -24,54 +35,107 @@ public IActionResult Index()
         [HttpPost]
         public IActionResult Index(LoginViewModel model)
         {
-            // TODO: HEEDLS-364 - Overwrite this old code for automatic sign in with new code that signs in the user after validation
-            if (model.Username?.ToLower() == "testusernameerror")
+            if (!ModelState.IsValid)
             {
-                ModelState.AddModelError("username", "There is a username error.");
+                return View("Index", model);
             }
 
-            if (model.Username?.ToLower() == "testpassworderror")
+            AdminUser? adminUser = userService.GetAdminUserByUsername(model.Username);
+            List<DelegateUser> delegateUsers = userService.GetDelegateUsersByUsername(model.Username);
+
+            if (adminUser == null && delegateUsers.Count == 0)
             {
-                ModelState.AddModelError("password", "There is a password error.");
+                ModelState.AddModelError("Username","No account with that email address or user ID could be found.");
+                return View("Index", model);
             }
 
-            if (!ModelState.IsValid)
+            adminUser = cryptoService.VerifyHashedPassword(adminUser?.Password, model.Password) ? adminUser : null;
+            delegateUsers = delegateUsers.Where(du => cryptoService.VerifyHashedPassword(du.Password, model.Password)).ToList();
+
+            if (adminUser == null && delegateUsers.Count == 0)
             {
+                ModelState.AddModelError("Password", "The password you have entered is incorrect.");
                 return View("Index", model);
             }
 
-            var claims = new List<Claim>
+            if (adminUser == null)

```

Comment:
All of this is quite long and could potentially be put into a service or a method. Also, we should probably follow this logic:
// Get first approved delegate
// If approvedDelegate == null && adminUser == null, then Go to account not approved page (note that if the user logs in with their delegate id and have an admin account associated with the same email address as the delegate account has, we should still go to the unapproved page, because they're trying to access the delegate account specifically and need communication about why their sign in isn't working)
// Else try and find an admin account associated with the approved & verified delegate account and use that to set some claims.